lib/SILOptimizer/Transforms/AccessEnforcementOpts.cpp

```diff
@@ -317,48 +319,55 @@ class AccessConflictAndMergeAnalysis {
   void
   detectApplyConflicts(const swift::FunctionAccessedStorage &callSiteAccesses,
-                       const DenseAccessMap &conflictFreeSet,
+                       const DenseAccessVec &conflictFreeSet,
                        const swift::FullApplySite &fullApply, RegionInfo &info);
+
+  void detectMayReleaseConflicts(const DenseAccessVec &conflictFreeSet,
+                                 SILInstruction *instr, RegionInfo &info);
 };
 } // namespace
 
 void AccessConflictAndMergeAnalysis::addInScopeAccess(
     RegionInfo &info, BeginAccessInst *beginAccess) {
-  auto &ai = result.getAccessInfo(beginAccess);
-  auto index = ai.getAccessIndex();
-  assert(info.getInScopeAccesses().find(index) ==
-             info.inScopeConflictFreeAccesses.conflictFreeAccesses.end() &&
-         "the begin_access should not have been in Set.");
-  info.inScopeConflictFreeAccesses.conflictFreeAccesses.insert(
-      std::make_pair(index, beginAccess));
+  assert(
+      std::find(info.inScopeConflictFreeAccesses.conflictFreeAccesses.begin(),
+                info.inScopeConflictFreeAccesses.conflictFreeAccesses.end(),
+                beginAccess) ==
+          info.inScopeConflictFreeAccesses.conflictFreeAccesses.end() &&
+      "the begin_access should not have been in Vec.");
+  info.inScopeConflictFreeAccesses.conflictFreeAccesses.insert(beginAccess);
 }
 
 void AccessConflictAndMergeAnalysis::removeInScopeAccess(
     RegionInfo &info, BeginAccessInst *beginAccess) {
-  auto &ai = result.getAccessInfo(beginAccess);
-  auto index = ai.getAccessIndex();
-  auto it = info.inScopeConflictFreeAccesses.conflictFreeAccesses.find(index);
+  auto it = std::find(
+      info.inScopeConflictFreeAccesses.conflictFreeAccesses.begin(),
+      info.inScopeConflictFreeAccesses.conflictFreeAccesses.end(), beginAccess);
   assert(it != info.inScopeConflictFreeAccesses.conflictFreeAccesses.end() &&
-         "the begin_access should have been in Set.");
+         "the begin_access should have been in Vec.");
   info.inScopeConflictFreeAccesses.conflictFreeAccesses.erase(it);
 }
 
 void AccessConflictAndMergeAnalysis::removeConflictFromStruct(
     RegionInfo &info, RegionInfo::AccessSummary &accessStruct,
-    const AccessedStorage &storage) {
-  auto pred = [&](const std::pair<unsigned, BeginAccessInst *> &pair) {
-    auto &currStorage = result.getAccessInfo(pair.second);
+    const AccessedStorage &storage, bool isInScope) {
+  auto pred = [&](BeginAccessInst *it) {
+    auto &currStorage = result.getAccessInfo(it);
     return !currStorage.isDistinctFrom(storage);
   };
   auto it = std::find_if(accessStruct.conflictFreeAccesses.begin(),
                          accessStruct.conflictFreeAccesses.end(), pred);
   while (it != accessStruct.conflictFreeAccesses.end()) {
+    if (isInScope) {
+      auto &ai = result.getAccessInfo(*it);
+      ai.setSeenNestedConflict();
+    }
     accessStruct.conflictFreeAccesses.erase(it);
```

Comment:
This still invalidates the iterator.

Reply:
Hmm yes: it might invalidate the iterator in `recordConflict`'s caller. not `it` in this function of course. I reverted to the old behavior in the caller